dynix.3.2.0/src/doc/man8/bad144.8
.\" $Copyright:	$
.\" Copyright (c) 1984, 1985, 1986, 1987, 1988, 1989, 1990 
.\" Sequent Computer Systems, Inc.   All rights reserved.
.\"  
.\" This software is furnished under a license and may be used
.\" only in accordance with the terms of that license and with the
.\" inclusion of the above copyright notice.   This software may not
.\" be provided or otherwise made available to, or used by, any
.\" other person.  No title to or ownership of the software is
.\" hereby transferred.
...
.V= $Header: bad144.8 1.8 86/05/13 $
.TH BAD144 8 "\*(V)" "4BSD"
.SH NAME
bad144 \- read/write dec standard 144 bad sector information
.SH SYNOPSIS
.B /etc/bad144
[
.B \-f
]
disktype disk
[ sno [
bad ...
] ]
.SH DESCRIPTION
.I Bad144
can be used to inspect the information stored on a disk that is used by
the disk drivers to implement bad sector forwarding.  The format of
the information is specified by DEC standard 144 (as extended by Sequent),
as follows.
.PP
The bad sector data is located in the last track of the disk.
There are five identical copies of each block of the data.
The first block, relative block 0,
is located in the first 5 even-numbered sectors of the track.
The second block (if any) is located in the first five odd-numbered sectors
of the track.
The third block (if any) is located in sectors 10-14 of the track.
Subsequent blocks (if any) are located in
subsequent five-block contiguous groups.
.PP
Replacement sectors are allocated starting with the first sector before
the bad sector information and working backwards towards the beginning
of the disk.
The position
of the bad sector in the bad sector table determines which replacement
sector it corresponds to.
The bad sectors must be listed in ascending order.
.PP
The bad sector information and replacement sectors are conventionally
only accessible through the ``c'' file system partition of the disk.  If
that partition is used for a file system, the user is responsible for
making sure that it does not overlap the bad sector information or any
replacement sectors.
.PP
The bad sector structure is described in /usr/sys/mbad/dkbad.h.
.PP
.I Bad144
is invoked by giving a device type (e.g. eagle), and a device
name (e.g. xp0, xp1, etc.).
It reads the first sector of the last track
of the corresponding disk and prints out the bad sector information.
It may also be invoked giving a serial number for the pack and a list
of bad sectors, and will then write the supplied information onto the
same location.  Note, however, that 
.I bad144
does not arrange for the specified sectors to be marked bad in this case.
This option should only be used to restore known bad sector information which
was destroyed.
It is necessary to reboot before the change will take effect.
.PP
If the disk is a Fujitsu Eagle, the
.B \-f
option may be used to mark the bad sectors as ``bad''.
\f3NOTE:  this can only be done safely when there is no other disk activity\fP,
preferably while running single-user.
Otherwise,
new bad sectors can be added only
by running a formatter.
Note that the order in which the sectors are listed determines which sectors
used for replacements; if new sectors are being inserted into the list on a
drive that is in use, care should be taken that replacements for
existing bad sectors have the correct contents.
.SH SEE ALSO
addbad(8),
badsect(8).
.SH BUGS
It should be possible to format disks on-line under UNIX.
.PP
It should be possible to mark bad sectors on drives of all types.
.PP
The drivers which write a system core image on disk after a crash do not
handle errors; thus the crash dump area must be free of errors and bad
sectors.


dynix.3.2.0/src/etc/bad144.c
/* $Copyright:	$
 * Copyright (c) 1984, 1985, 1986, 1987, 1988, 1989, 1990 
 * Sequent Computer Systems, Inc.   All rights reserved.
 *  
 * This software is furnished under a license and may be used
 * only in accordance with the terms of that license and with the
 * inclusion of the above copyright notice.   This software may not
 * be provided or otherwise made available to, or used by, any
 * other person.  No title to or ownership of the software is
 * hereby transferred.
 */

#ifndef lint
static char rcsid[] = "$Header: bad144.c 2.1 86/05/18 $";
#endif

/*
 * bad144
 *
 * This program prints and/or initializes a bad block record for a pack,
 * in the format used by the DEC standard 144 as modified by Sequent.
 *
 * It is preferable to write the bad information with a standard formatter,
 * but this program will do in a pinch, e.g. if the bad information is
 * accidentally wiped out this is a much faster way of restoring it than
 * reformatting. 
 */

#include <sys/types.h>
#include <mbad/dkbad.h>
#include <sys/ioctl.h>
#include <sys/file.h>
#include <mbad/xpioctl.h>

#include <stdio.h>
#include <disktab.h>

int	fflag;
struct  dkbad dkbad[DK_NBADMAX];

main(argc, argv)
	int argc;
	char *argv[];
{
	register union bt_bad *bt;
	register struct disktab *dp;
	char name[BUFSIZ];
	int size, i, f, bad, errs;
	int block, blks;

	argc--, argv++;
	if (argc > 0 && strcmp(*argv, "-f") == 0) {
		argc--, argv++;
		fflag++;
	}
	if (argc < 2) {
		fprintf(stderr,
		  "usage: bad144 [ -f ] type disk [ snum [ bn ... ] ]\n");
		fprintf(stderr, "e.g.: bad144 eagle xp0\n");
		exit(1);
	}
	dp = getdiskbyname(argv[0]);
	if (dp == NULL) {
		fprintf(stderr, "%s: unknown disk type\n", argv[0]);
		exit(1);
	}
	if (strcmp(dp->d_fmt, "dec144") != 0) {
		fprintf(stderr, "%s: not dec144 disk format.\n", dp->d_name);
		exit(1);
	}
	sprintf(name, "/dev/r%sc", argv[1]);
	argc -= 2;
	argv += 2;
	size = dp->d_nsectors * dp->d_ntracks * dp->d_ncylinders; 
	if (argc == 0) {
		bt = dkbad[0].bt_bad;
		for(i=0; i<DK_MAXBAD; i++)
			bt[i].bt_cyl = bt[i].bt_trksec = DK_END;
		f = open(name, O_RDONLY);
		if (f < 0)
			Perror(name);
		printf("bad block information at sector %d in %s:\n",
		    			size - dp->d_nsectors, name);
		block = 0;
		blks = 1;
		readbad(f, dp, size, name, block, blks);
		if((blks = dkbad[0].bt_lastb) > 0) {
			block++;
			readbad(f, dp, size, name, block, blks);
		}
		printf("cartidge serial number: %d(10)\n", dkbad[0].bt_csn);
		switch (dkbad[0].bt_flag) {

		case -1:
			printf("alignment cartridge\n");
			break;

		case 0:
			break;

		default:
			printf("bt_flag=%x(16)?\n", dkbad[0].bt_flag);
			break;
		}
		for (i = 0; i < DK_MAXBAD; i++) {
			bad = (bt[i].bt_cyl<<16) + bt[i].bt_trksec;
			if (bad < 0)
				break;
			printf("sn=%d, cn=%d, tn=%d, sn=%d\n",
			    (bt[i].bt_cyl*dp->d_ntracks + (bt[i].bt_trksec>>8))
				 * dp->d_nsectors + (bt[i].bt_trksec&0xff),
			    bt[i].bt_cyl, bt[i].bt_trksec>>8, bt[i].bt_trksec&0xff);
		}
		exit(0);
	}
	f = open(name, 1 + fflag);
	if (f < 0)
		Perror(name);
	dkbad[0].bt_csn = atoi(*argv++);
	bt = dkbad[0].bt_bad;
	for(i=0; i<DK_MAXBAD; i++)
		bt[i].bt_cyl = bt[i].bt_trksec = DK_END;
	argc--;
	if (argc > DK_MAXBAD) {
		printf("bad144: too many bad sectors specified\n");
		printf("limited to %d by information format\n", DK_MAXBAD);
		exit(1);
	}
	errs = 0;
	i = 0;
	while (argc > 0) {
		int sn = atoi(*argv++);

		argc--;
		if (sn <= 0 || sn >= size) {
			printf("%d: out of range 1 to %d for %s\n",
			    sn, size - 1, dp->d_name);
			errs++;
		}
		bt[i].bt_cyl = sn / (dp->d_nsectors*dp->d_ntracks);
		sn %= (dp->d_nsectors*dp->d_ntracks);
		bt[i].bt_trksec =
		    ((sn/dp->d_nsectors) << 8) + (sn%dp->d_nsectors);
		i++;
	}
	if (errs)
		exit(1);

	if(i <= DK_NBAD_0)
		dkbad[0].bt_lastb = 0;
	else
		dkbad[0].bt_lastb = ((i - DK_NBAD_0)/DK_NBAD_N) + 1;

	writebad(f, dp, size, name);

	if (fflag)
		for (i = 0; i < DK_MAXBAD; i++) {
			daddr_t bn;

			bad = (bt[i].bt_cyl<<16) + bt[i].bt_trksec;
			if (bad < 0)
				break;
			bn = (bt[i].bt_cyl * dp->d_ntracks +
			    (bt[i].bt_trksec >> 8)) *
			    dp->d_nsectors + (bt[i].bt_trksec & 0xff);
			format(f, dp, bn);
		}
	exit(0);
}

/*
 * read bad block list
 */

readbad(f, dp, size, name, block, blks)
	int f, size, block;
	register int blks;
	struct disktab *dp;
	char *name;
{
	register int copy;

	do {
		for(copy=0; copy<DK_NBADCOPY; copy++) {
			if(lseek(f, dp->d_secsize * (size - dp->d_nsectors + 
					DK_LOC(block, copy)), L_SET) < 0)
				Perror("lseek");
			if(read(f, (char *)&dkbad[block], sizeof(struct dkbad))
					== sizeof(struct dkbad))
				break;
		}
		if(copy == DK_NBADCOPY) {
			fprintf(stderr, "bad144: %s: can't read bad block info\n", name);
				exit(1);
		}
		block++;
	} while(--blks);
}

/*
 * write the bad block list
 */

writebad(f, dp, size, name)
	int f, size;
	struct disktab *dp;
	char *name;
{
	int block, copy;

	block = 0;
	do {
		for(copy=0; copy<DK_NBADCOPY; copy++) {
			if(lseek(f, dp->d_secsize * (size - dp->d_nsectors + 
					DK_LOC(block, copy)), L_SET) < 0) {
				fprintf(stderr, "bad144: %s: can't write bad block %d copy %d, seek failed\n", name, block, copy);
				continue;
			}
			if(write(f, (char *)&dkbad[block], sizeof(struct dkbad))
					!= sizeof(struct dkbad))
				fprintf(stderr, "bad144: %s: can't write bad block %d copy %d\n", name, block, copy);
		}
	} while(++block < DK_NBADMAX);
}

/*
 *	Routine to zap a header on xp
 */

zapxp(fd, dp, blk)
	int fd;
	struct disktab *dp;
	daddr_t blk;
{
	struct xp_ioctl s;
	daddr_t t;

	s.x_cyl = blk / (dp->d_nsectors*dp->d_ntracks);
	t = blk % (dp->d_nsectors*dp->d_ntracks);
	s.x_head = t / dp->d_nsectors;
	s.x_sector = t % dp->d_nsectors;
	/* Allow for adaptive format */
	s.x_sector = (s.x_sector + s.x_head) % dp->d_nsectors;

	if (ioctl(fd, XPIOCZAPHEADER, &s) < 0)
		fprintf(stderr, "bad144: can't mark sector %d\n", blk);
}


struct	formats {
	char	*f_name;		/* disk name */
	int	(*f_routine)();		/* routine for special handling */
} formats[] = {
	{ "eagle", zapxp },
	{ 0, 0 }
};

format(fd, dp, blk)
	int fd;
	struct disktab *dp;
	daddr_t blk;
{
	register struct formats *fp;

	for (fp = formats; fp->f_name; fp++)
		if (strcmp(dp->d_name, fp->f_name) == 0)
			break;
	if (fp->f_name == 0) {
		fprintf(stderr, "bad144: don't know how to format %s disks\n",
			dp->d_name);
		exit(2);
	}
	(*fp->f_routine)(fd, dp, blk);
}

Perror(op)
	char *op;
{

	fprintf(stderr, "bad144: "); perror(op);
	exit(4);
}
